h/accounts/views.py

```diff
@@ -495,34 +495,49 @@ def get(self):
         """Show the user's account."""
         return self._template_data()
 
-    @view_config(request_method='POST')
-    def post(self):
-        """Handle POST payload from account update form."""
-        formid = self.request.POST.get('__formid__')
-        if formid is None or formid not in self.forms:
-            raise httpexceptions.HTTPBadRequest()
+    @view_config(request_method='POST',
+                 request_param='__formid__=email')
+    def post_email_form(self):
+        """Called by Pyramid when the change email form is submitted."""
+        return self._post(self._handle_email_form)
+
+    @view_config(request_method='POST',
+                 request_param='__formid__=password')
+    def post_password_form(self):
+        """Called by Pyramid when the change password form is submitted."""
+        return self._post(self._handle_password_form)
 
+    def _handle_email_form(self):
         try:
-            if formid == 'email':
-                self._handle_email_form()
-            elif formid == 'password':
-                self._handle_password_form()
+            appstruct = self.forms['email'].validate(self.request.POST.items())
+        except deform.ValidationFailure:
+            raise
```

Comment:
I think I understand why you've done this, but to me this just adds unnecessary noise to understanding what this function does.

One of the nicest features of Pyramid (IMO) is the exception view, which allows you to move error handling out of the main flow of the code, so that understanding what the code is supposed to do in the "happy path" case is as easy as possible. You almost seem to be going down the opposite route here...

In general, I think something like this is what we should be shooting for:

```python
class FormController(object):

    def __init__(self, context, request):
        self.context = context
        self.request = request
        self.form = request.create_form(...)

    @view_config(request_method='GET')
    def get(self):
        return {'form': self.form.render()}

    @view_config(request_method='POST')
    def post(self):
        appstruct = self.form.validate(self.request.POST.items())

        # do stuff with appstruct

        return HTTPFound(location=request.path)

    @view_config(context=deform.ValidationFailure)
    def validation_failure(self):
        return {'form': self.context.render()}
```

The important point here is not that I'm using an exception view per se, but that the error-handling code is a) generic (it will work for any views in this class) and b) distinct from the view code itself, making both the happy path and the error handling code easier to test.

In this particular context (the account controller) it's made a little more complicated by a variety of factors (multiple forms, additional context, and so on), but whether it's an exception view, a decorator, or even just a helper function, we should be trying to make it possible to write correct form handlers across the board, whereas I'm worried that these changes just make it harder to understand what's happening, and are closely tied to this particular controller.

Perhaps an even simpler approach would be a `validate_form` helper function, something like:

```python
class validate_form(request, form):
    try:
        appstruct = form.validate(request.POST.items())
    except deform.ValidationFailure:
        return None, {'form': form.render()}
    else:
        return appstruct, HTTPFound(location=request.path)
```

which you could then use as:

```
@view_config(request_method='POST', request_param='__formid__=email')
def post_email(self):
    appstruct, response = validate_form(self.request, self.forms['email'])
    if appstruct is not None:  # form was valid
        # update email
    else:
        response.update({'email': request.authenticated_user.email})
    return response
```

This relies on `response` having different types when the form is valid vs invalid, but I don't think that's so bad...